### PR TITLE
feat(cfn): Append hash to outputs

### PR DESCRIPTION
**Pull request by LaikaN57**
This error appears sometimes when editing a cfn template:

> 00:39:22 CRITICAL [DRY: S3-LOGGING-BUCKET (us-east-2)] Unexpected Change Set state (Status) received (FAILED): No updates are to be performed.

This is a bit of an oddity between `kingpin` and CFN.

`kingpin` is pushing an update to CFN because the template has changed, but CFN is denying that there are any changes because physical resources in CFN template have not actually changes. This is likely because the PR is only making changes to conditional resources in the CFN template yet there are CFN stacks using this template that do not deploy that conditional resource - therefore leading to CFN believing there was "no update".

~~To resolve this, they need to bump the metadata in the CFN template on a non-conditional resource.~~

^ This action is no longer required of the engineer.

Through testing, we have figured out that modifying an `Outputs` map entry counts as a CFN change. Therefore we have added some code to automatically append a special output entry that `kingpin` will manage internally. If there is a diff in the live stack template vs the template we input into the `kingpin` script, our new code will ensure we add an extra output value hash when creating a changeset. We also take care to strip this managed output when re-downloading the stack from the live state.

You can disable all of this functionality by setting the `KINGPIN_CFN_HASH_OUTPUT_KEY` env var to an empty string.

### CODE REVIEW

File: kingpin/actors/aws/cloudformation.py [hunk 1]
```diff
@@ -230,6 +232,33 @@ def _discover_default_params(self, template_body):
...
+    def _strip_hash(self, template_str: str, as_obj: bool = False):
+        """Strips the hash from the template.
+
+        Note: This will also strip the "Outputs" section if no other output exists. This might cause
+        issues when diffiing a template that contains an outputs section with no outputs.
+        """
+
+        template_obj = json.loads(template_str)
+
+        if len(KINGPIN_CFN_HASH_OUTPUT_KEY) > 0:
```
Comment (diranged) [inline kingpin/actors/aws/cloudformation.py:244]: 1. Why are we doing a length check? Why not:
```python
if KINGPIN_CFN_HASH_OUTPUT_KEY:
  ...
```

2. It's a bit more "pythonic" to do something like:
```python
if !KINGPIN_CFN_HASH_OUTPUT_KEY:
  return template_str
``

File: kingpin/actors/aws/cloudformation.py [hunk 2]
```diff
@@ -230,6 +232,33 @@ def _discover_default_params(self, template_body):
         }
         return default_params
 
+    def _strip_hash(self, template_str: str, as_obj: bool = False):
```
Comment (diranged) [inline kingpin/actors/aws/cloudformation.py:235]: You have no return type here - yet you return stuff.

File: kingpin/actors/aws/cloudformation.py [hunk 3]
```diff
@@ -230,6 +232,33 @@ def _discover_default_params(self, template_body):
...
+        """Strips the hash from the template.
+
+        Note: This will also strip the "Outputs" section if no other output exists. This might cause
+        issues when diffiing a template that contains an outputs section with no outputs.
+        """
+
+        template_obj = json.loads(template_str)
+
+        if len(KINGPIN_CFN_HASH_OUTPUT_KEY) > 0:
+            if "Outputs" not in template_obj.keys() or not isinstance(
```
Comment (diranged) [inline kingpin/actors/aws/cloudformation.py:245]: I'm confused by this check. Wouldn't this quit the function and return if someone had a simple template that just had no outputs? Wouldn't we _want to add the outputs section_ in that case?

File: kingpin/actors/aws/cloudformation.py [hunk 4]
```diff
@@ -230,6 +232,33 @@ def _discover_default_params(self, template_body):
...
+                template_obj["Outputs"], dict
+            ):
+                return template_str
+
+            template_outputs_keys = template_obj["Outputs"].keys()
+            if KINGPIN_CFN_HASH_OUTPUT_KEY in template_outputs_keys:
+                del template_obj["Outputs"][KINGPIN_CFN_HASH_OUTPUT_KEY]
+
+                if len(template_outputs_keys) == 0:
```
Comment (diranged) [inline kingpin/actors/aws/cloudformation.py:254]: Why is this necessary?

File: kingpin/actors/aws/cloudformation.py [hunk 5]
```diff
@@ -230,6 +232,33 @@ def _discover_default_params(self, template_body):
...
+                    del template_obj["Outputs"]
+
+        if as_obj:
+            return template_obj
```
Comment (diranged) [inline kingpin/actors/aws/cloudformation.py:258]: Don't do this inside the function. Make one function that returns the object, and then another function that calls the object-function and then stringifies it ... if necessary.

File: kingpin/actors/aws/cloudformation.py [hunk 6]
```diff
@@ -1080,6 +1112,23 @@ def _diff_params_safely(self, remote, local):
...
+
+        template_obj = json.loads(self._template_body)
+
+        if not isinstance(template_obj.get("Outputs", None), dict):
+            # overwrite the outputs with an empty dict
+            template_obj["Outputs"] = {}
+
+        template_obj["Outputs"][KINGPIN_CFN_HASH_OUTPUT_KEY] = {
+            "Value": md5(
+                datetime.datetime.now(datetime.timezone.utc).isoformat().encode()
```
Comment (diranged) [inline kingpin/actors/aws/cloudformation.py:1126]: What are we doing here? This value will change on every deploy. I thought you were going to hash the template contents?

File: kingpin/actors/aws/cloudformation.py (hunk 3 above)
Comment (LaikaN57) [reply kingpin/actors/aws/cloudformation.py]: I think I answered these two comments in standup but LMK if these were about something different. There are 2 phases that kingpin will go through. A diff phase and a push phase. In the diff phase kingpin will collect info from the local repo about what that stack template should be and also collect info from the AWS CFN API to see what the live stack template is at the present moment. In order to remove complication, kingpin will silently strip out this managed hash value before the diff phase. However, in order to avoid the PR description issue, we will add a new hash value back to the template that we push in the changeset during the push phase. (... and in the next update we will again need to strip the hash for diff'ing, etc.)

File: kingpin/actors/aws/cloudformation.py (hunk 4 above)
Comment (LaikaN57) [reply kingpin/actors/aws/cloudformation.py]: I think I answered these two comments in standup but LMK if these were about something different. There are 2 phases that kingpin will go through. A diff phase and a push phase. In the diff phase kingpin will collect info from the local repo about what that stack template should be and also collect info from the AWS CFN API to see what the live stack template is at the present moment. In order to remove complication, kingpin will silently strip out this managed hash value before the diff phase. However, in order to avoid the PR description issue, we will add a new hash value back to the template that we push in the changeset during the push phase. (... and in the next update we will again need to strip the hash for diff'ing, etc.)

File: kingpin/actors/aws/cloudformation.py (hunk 1 above)
Comment (LaikaN57) [reply kingpin/actors/aws/cloudformation.py]: I will test how python handles env vars and the difference between passing in an empty string vs a "null" (in bash). I thought it would be easier to just keep it as a string and do a length comparison. That way a user could disable the hash-management functionality by passing an empty string where as the default case would be that we do use the hash functionality.

File: kingpin/actors/aws/cloudformation.py [hunk 7]
```diff
@@ -1080,6 +1114,29 @@ def _diff_params_safely(self, remote, local):
 
         return False
 
+    def _template_body_with_hash(self) -> str:
+        """Add a hash to the template to force a change in the stack."""
+
+        # Bail if the user has disabled this feature.
+        if len(KINGPIN_CFN_HASH_OUTPUT_KEY) == 0:
```
[the remark this answers is not in the document]
Comment (LaikaN57) [reply kingpin/actors/aws/cloudformation.py:1121]: I will test how python handles env vars and the difference between passing in an empty string vs a "null" (in bash). I thought it would be easier to just keep it as a string and do a length comparison. That way a user could disable the hash-management functionality by passing an empty string where as the default case would be that we do use the hash functionality.

[the remark this answers is not in the document]
Comment (LaikaN57) [reply kingpin/actors/aws/cloudformation.py:1121]: My intent is the following:

If env var is:
- a string of length > 0, use it as the output key
- an empty string, disable functionality
- not set or otherwise null, use the default output key

[the remark this answers is not in the document]
Comment (LaikaN57) [reply kingpin/actors/aws/cloudformation.py:1121]: Using this simple script here are the results of what I found...

```python
import os

ADK_CONST = os.getenv("ADK_CONST", "adkdefault")

if __name__ == "__main__":
    print(f"ADK_CONST: \"{ADK_CONST}\"")
    print(f"ADK_CONST type: \"{type(ADK_CONST)}\"")
```

```
.venv) akennedy@ndm2a kingpin % python adk.py              
ADK_CONST: "adkdefault"
ADK_CONST type: "<class 'str'>"
(.venv) akennedy@ndm2a kingpin % env ADK_CONST= python adk.py
ADK_CONST: ""
ADK_CONST type: "<class 'str'>"
(.venv) akennedy@ndm2a kingpin % ADK_CONST= python adk.py 
ADK_CONST: ""
ADK_CONST type: "<class 'str'>"
(.venv) akennedy@ndm2a kingpin % env ADK_CONST="" python adk.py
ADK_CONST: ""
ADK_CONST type: "<class 'str'>"
(.venv) akennedy@ndm2a kingpin % ADK_CONST="" python adk.py    
ADK_CONST: ""
ADK_CONST type: "<class 'str'>"
(.venv) akennedy@ndm2a kingpin % env ADK_CONST="adkoverride" python adk.py
ADK_CONST: "adkoverride"
ADK_CONST type: "<class 'str'>"
(.venv) akennedy@ndm2a kingpin % ADK_CONST="adkoverride" python adk.py 
ADK_CONST: "adkoverride"
ADK_CONST type: "<class 'str'>"
```

[the remark this answers is not in the document]
Comment (diranged) [reply kingpin/actors/aws/cloudformation.py:1121]: But ... you don't need to do that?
```python
>>> import os
>>> def isTrue(str: str):
...     if (str):
...         print("true")
...     else:
...         print("false")
...         
>>> isTrue(os.getenv("MISSING"))
false
>>> isTrue(os.getenv("USER"))
true
>>> 
```

[the remark this answers is not in the document]
Comment (diranged) [reply kingpin/actors/aws/cloudformation.py:1121]: so simply writing: `if KINGPIN_CFN_HASH_OUTPUT_KEY: ...` should work.

**Review by diranged — DISMISSED** [review]

**Comment by LaikaN57** [comment]
e2e testing:

```yaml
# stack.yaml
Conditions:
  FalseA: !Equals ["a", "b"]

Resources:
  MyTopic:
    Type: AWS::SNS::Topic

  MyTopicConditional:
    Condition: FalseA
    Type: AWS::SNS::Topic
    Properties:
      TopicName: "AkennedyTest"
```

Run 1 - creates that stack, no hashing required.

```
(.venv) akennedy@ndm2a kingpin % kingpin -j ./deploy.yaml
akennedy - kingpin/bin/deploy.py
2024-11-27 14:19:00,829 INFO      Rehearsing... Break a leg!
2024-11-27 14:19:00,829 INFO      [DRY: Kingpin] Preparing actors from ./deploy.yaml
2024-11-27 14:19:00,991 INFO      [DRY: kingpin.actors.group.Sync] Beginning 1 actions
2024-11-27 14:19:00,991 INFO      [DRY: CloudFormation Stack akennedy-test-kingpin-1] Validating template with AWS...
2024-11-27 14:19:01,812 INFO      [DRY: CloudFormation Stack akennedy-test-kingpin-1] Ensuring that CF Stack akennedy-test-kingpin-1 is present
2024-11-27 14:19:01,878 WARNING   [DRY: CloudFormation Stack akennedy-test-kingpin-1] Would have created stack akennedy-test-kingpin-1
2024-11-27 14:19:01,879 INFO      Rehearsal OK! Performing!
2024-11-27 14:19:01,879 INFO      [Kingpin] Preparing actors from ./deploy.yaml
2024-11-27 14:19:01,903 INFO      
2024-11-27 14:19:01,903 WARNING   Lights, camera ... action!
2024-11-27 14:19:01,903 INFO      
2024-11-27 14:19:01,903 INFO      [kingpin.actors.group.Sync] Beginning 1 actions
2024-11-27 14:19:01,903 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Validating template with AWS...
2024-11-27 14:19:02,137 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Ensuring that CF Stack akennedy-test-kingpin-1 is present
2024-11-27 14:19:02,214 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Creating stack akennedy-test-kingpin-1
2024-11-27 14:19:02,619 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Stack state is CREATE_IN_PROGRESS, waiting 15(s)...
2024-11-27 14:19:17,832 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Stack created: arn:aws:cloudformation:us-west-2:000000000000:stack/akennedy-test-kingpin-1/9b486dd0-ad0d-11ef-ad77-02278992b85d
```

image


Run 2 - update "non-change" stack, add new hash

```diff
-  FalseA: !Equals ["a", "b"]
+  FalseA: !Equals ["a", "c"]
```

```
(.venv) akennedy@ndm2a kingpin % kingpin -j ./deploy.yaml
akennedy - kingpin/bin/deploy.py
2024-11-27 14:22:25,608 INFO      Rehearsing... Break a leg!
2024-11-27 14:22:25,608 INFO      [DRY: Kingpin] Preparing actors from ./deploy.yaml
2024-11-27 14:22:25,771 INFO      [DRY: kingpin.actors.group.Sync] Beginning 1 actions
2024-11-27 14:22:25,771 INFO      [DRY: CloudFormation Stack akennedy-test-kingpin-1] Validating template with AWS...
2024-11-27 14:22:26,573 INFO      [DRY: CloudFormation Stack akennedy-test-kingpin-1] Ensuring that CF Stack akennedy-test-kingpin-1 is present
2024-11-27 14:22:26,716 INFO      [DRY: CloudFormation Stack akennedy-test-kingpin-1] Verifying that stack is in desired state
2024-11-27 14:22:26,773 WARNING   [DRY: CloudFormation Stack akennedy-test-kingpin-1] Stack templates do not match.
2024-11-27 14:22:26,774 INFO      [DRY: CloudFormation Stack akennedy-test-kingpin-1] Diff: --- 
2024-11-27 14:22:26,774 INFO      [DRY: CloudFormation Stack akennedy-test-kingpin-1] Diff: 
2024-11-27 14:22:26,774 INFO      [DRY: CloudFormation Stack akennedy-test-kingpin-1] Diff: +++ 
2024-11-27 14:22:26,774 INFO      [DRY: CloudFormation Stack akennedy-test-kingpin-1] Diff: 
2024-11-27 14:22:26,774 INFO      [DRY: CloudFormation Stack akennedy-test-kingpin-1] Diff: @@ -1,3 +1,3 @@
2024-11-27 14:22:26,774 INFO      [DRY: CloudFormation Stack akennedy-test-kingpin-1] Diff: 
2024-11-27 14:22:26,774 INFO      [DRY: CloudFormation Stack akennedy-test-kingpin-1] Diff: -[('Conditions', [('FalseA', [('Fn::Equals', ['a', 'b'])])]),
2024-11-27 14:22:26,774 INFO      [DRY: CloudFormation Stack akennedy-test-kingpin-1] Diff: +[('Conditions', [('FalseA', [('Fn::Equals', ['a', 'c'])])]),
2024-11-27 14:22:26,774 INFO      [DRY: CloudFormation Stack akennedy-test-kingpin-1] Diff:   ('Resources',
2024-11-27 14:22:26,774 INFO      [DRY: CloudFormation Stack akennedy-test-kingpin-1] Diff:    [('MyTopic', [('Type', 'AWS::SNS::Topic')]),
2024-11-27 14:22:26,774 INFO      [DRY: CloudFormation Stack akennedy-test-kingpin-1] Generating a stack Change Set...
2024-11-27 14:22:26,930 INFO      [DRY: CloudFormation Stack akennedy-test-kingpin-1] Waiting for arn:aws:cloudformation:us-west-2:000000000000:changeSet/kingpin-e0aadc7096244c0e8e4a94a3c66d49d1/e01b40d8-b21d-4dc1-bb75-f28a04a1da43 to reach CREATE_COMPLETE
2024-11-27 14:22:27,019 INFO      [DRY: CloudFormation Stack akennedy-test-kingpin-1] Change Set state is CREATE_PENDING, waiting 5(s)...
2024-11-27 14:22:32,111 WARNING   [DRY: CloudFormation Stack akennedy-test-kingpin-1] Would have executed Change Set arn:aws:cloudformation:us-west-2:000000000000:changeSet/kingpin-e0aadc7096244c0e8e4a94a3c66d49d1/e01b40d8-b21d-4dc1-bb75-f28a04a1da43
2024-11-27 14:22:32,206 INFO      [DRY: CloudFormation Stack akennedy-test-kingpin-1] Done updating template
2024-11-27 14:22:32,207 INFO      Rehearsal OK! Performing!
2024-11-27 14:22:32,208 INFO      [Kingpin] Preparing actors from ./deploy.yaml
2024-11-27 14:22:32,255 INFO      
2024-11-27 14:22:32,255 WARNING   Lights, camera ... action!
2024-11-27 14:22:32,255 INFO      
2024-11-27 14:22:32,255 INFO      [kingpin.actors.group.Sync] Beginning 1 actions
2024-11-27 14:22:32,255 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Validating template with AWS...
2024-11-27 14:22:32,464 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Ensuring that CF Stack akennedy-test-kingpin-1 is present
2024-11-27 14:22:32,601 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Verifying that stack is in desired state
2024-11-27 14:22:32,658 WARNING   [CloudFormation Stack akennedy-test-kingpin-1] Stack templates do not match.
2024-11-27 14:22:32,658 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Diff: --- 
2024-11-27 14:22:32,658 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Diff: 
2024-11-27 14:22:32,658 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Diff: +++ 
2024-11-27 14:22:32,659 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Diff: 
2024-11-27 14:22:32,659 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Diff: @@ -1,3 +1,3 @@
2024-11-27 14:22:32,659 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Diff: 
2024-11-27 14:22:32,659 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Diff: -[('Conditions', [('FalseA', [('Fn::Equals', ['a', 'b'])])]),
2024-11-27 14:22:32,659 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Diff: +[('Conditions', [('FalseA', [('Fn::Equals', ['a', 'c'])])]),
2024-11-27 14:22:32,659 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Diff:   ('Resources',
2024-11-27 14:22:32,659 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Diff:    [('MyTopic', [('Type', 'AWS::SNS::Topic')]),
2024-11-27 14:22:32,659 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Generating a stack Change Set...
2024-11-27 14:22:32,848 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Waiting for arn:aws:cloudformation:us-west-2:000000000000:changeSet/kingpin-e0aadc7096244c0e8e4a94a3c66d49d1/e346d134-494e-4548-8b0b-0f247b15665a to reach CREATE_COMPLETE
2024-11-27 14:22:32,928 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Change Set state is CREATE_PENDING, waiting 5(s)...
2024-11-27 14:22:38,015 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Executing change set arn:aws:cloudformation:us-west-2:000000000000:changeSet/kingpin-e0aadc7096244c0e8e4a94a3c66d49d1/e346d134-494e-4548-8b0b-0f247b15665a
2024-11-27 14:22:38,188 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Waiting for arn:aws:cloudformation:us-west-2:000000000000:changeSet/kingpin-e0aadc7096244c0e8e4a94a3c66d49d1/e346d134-494e-4548-8b0b-0f247b15665a to reach EXECUTE_COMPLETE
2024-11-27 14:22:38,258 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Change Set state is AVAILABLE, waiting 5(s)...
2024-11-27 14:22:43,404 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Done updating template
```

image

Run 3 - update hash, also add an non-managed output

```
(.venv) akennedy@ndm2a kingpin % kingpin -j ./deploy.yaml
akennedy - kingpin/bin/deploy.py
2024-11-27 14:25:07,307 INFO      Rehearsing... Break a leg!
2024-11-27 14:25:07,308 INFO      [DRY: Kingpin] Preparing actors from ./deploy.yaml
2024-11-27 14:25:07,468 INFO      [DRY: kingpin.actors.group.Sync] Beginning 1 actions
2024-11-27 14:25:07,469 INFO      [DRY: CloudFormation Stack akennedy-test-kingpin-1] Validating template with AWS...
2024-11-27 14:25:08,306 INFO      [DRY: CloudFormation Stack akennedy-test-kingpin-1] Ensuring that CF Stack akennedy-test-kingpin-1 is present
2024-11-27 14:25:08,424 INFO      [DRY: CloudFormation Stack akennedy-test-kingpin-1] Verifying that stack is in desired state
2024-11-27 14:25:08,476 WARNING   [DRY: CloudFormation Stack akennedy-test-kingpin-1] Stack templates do not match.
2024-11-27 14:25:08,476 INFO      [DRY: CloudFormation Stack akennedy-test-kingpin-1] Diff: --- 
2024-11-27 14:25:08,476 INFO      [DRY: CloudFormation Stack akennedy-test-kingpin-1] Diff: 
2024-11-27 14:25:08,476 INFO      [DRY: CloudFormation Stack akennedy-test-kingpin-1] Diff: +++ 
2024-11-27 14:25:08,476 INFO      [DRY: CloudFormation Stack akennedy-test-kingpin-1] Diff: 
2024-11-27 14:25:08,476 INFO      [DRY: CloudFormation Stack akennedy-test-kingpin-1] Diff: @@ -1,3 +1,4 @@
2024-11-27 14:25:08,476 INFO      [DRY: CloudFormation Stack akennedy-test-kingpin-1] Diff: 
2024-11-27 14:25:08,476 INFO      [DRY: CloudFormation Stack akennedy-test-kingpin-1] Diff:  [('Conditions', [('FalseA', [('Fn::Equals', ['a', 'c'])])]),
2024-11-27 14:25:08,476 INFO      [DRY: CloudFormation Stack akennedy-test-kingpin-1] Diff: + ('Outputs', [('MyOutput', [('Value', 'MyOutputValue')])]),
2024-11-27 14:25:08,476 INFO      [DRY: CloudFormation Stack akennedy-test-kingpin-1] Diff:   ('Resources',
2024-11-27 14:25:08,476 INFO      [DRY: CloudFormation Stack akennedy-test-kingpin-1] Diff:    [('MyTopic', [('Type', 'AWS::SNS::Topic')]),
2024-11-27 14:25:08,476 INFO      [DRY: CloudFormation Stack akennedy-test-kingpin-1] Generating a stack Change Set...
2024-11-27 14:25:08,605 INFO      [DRY: CloudFormation Stack akennedy-test-kingpin-1] Waiting for arn:aws:cloudformation:us-west-2:000000000000:changeSet/kingpin-2192477344484497b06c1a7ee9c11f38/dec145d4-fbd9-40d1-bbf9-c091f071fb0c to reach CREATE_COMPLETE
2024-11-27 14:25:08,680 INFO      [DRY: CloudFormation Stack akennedy-test-kingpin-1] Change Set state is CREATE_PENDING, waiting 5(s)...
2024-11-27 14:25:13,776 WARNING   [DRY: CloudFormation Stack akennedy-test-kingpin-1] Would have executed Change Set arn:aws:cloudformation:us-west-2:000000000000:changeSet/kingpin-2192477344484497b06c1a7ee9c11f38/dec145d4-fbd9-40d1-bbf9-c091f071fb0c
2024-11-27 14:25:13,875 INFO      [DRY: CloudFormation Stack akennedy-test-kingpin-1] Done updating template
2024-11-27 14:25:13,875 INFO      Rehearsal OK! Performing!
2024-11-27 14:25:13,876 INFO      [Kingpin] Preparing actors from ./deploy.yaml
2024-11-27 14:25:13,912 INFO      
2024-11-27 14:25:13,912 WARNING   Lights, camera ... action!
2024-11-27 14:25:13,912 INFO      
2024-11-27 14:25:13,912 INFO      [kingpin.actors.group.Sync] Beginning 1 actions
2024-11-27 14:25:13,912 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Validating template with AWS...
2024-11-27 14:25:14,103 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Ensuring that CF Stack akennedy-test-kingpin-1 is present
2024-11-27 14:25:14,219 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Verifying that stack is in desired state
2024-11-27 14:25:14,269 WARNING   [CloudFormation Stack akennedy-test-kingpin-1] Stack templates do not match.
2024-11-27 14:25:14,269 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Diff: --- 
2024-11-27 14:25:14,270 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Diff: 
2024-11-27 14:25:14,270 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Diff: +++ 
2024-11-27 14:25:14,270 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Diff: 
2024-11-27 14:25:14,270 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Diff: @@ -1,3 +1,4 @@
2024-11-27 14:25:14,270 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Diff: 
2024-11-27 14:25:14,270 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Diff:  [('Conditions', [('FalseA', [('Fn::Equals', ['a', 'c'])])]),
2024-11-27 14:25:14,270 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Diff: + ('Outputs', [('MyOutput', [('Value', 'MyOutputValue')])]),
2024-11-27 14:25:14,270 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Diff:   ('Resources',
2024-11-27 14:25:14,270 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Diff:    [('MyTopic', [('Type', 'AWS::SNS::Topic')]),
2024-11-27 14:25:14,270 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Generating a stack Change Set...
2024-11-27 14:25:14,439 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Waiting for arn:aws:cloudformation:us-west-2:000000000000:changeSet/kingpin-2192477344484497b06c1a7ee9c11f38/45f15ea9-c4d4-4d60-94b9-9f0584a9e9fa to reach CREATE_COMPLETE
2024-11-27 14:25:14,516 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Change Set state is CREATE_PENDING, waiting 5(s)...
2024-11-27 14:25:19,599 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Executing change set arn:aws:cloudformation:us-west-2:000000000000:changeSet/kingpin-2192477344484497b06c1a7ee9c11f38/45f15ea9-c4d4-4d60-94b9-9f0584a9e9fa
2024-11-27 14:25:19,770 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Waiting for arn:aws:cloudformation:us-west-2:000000000000:changeSet/kingpin-2192477344484497b06c1a7ee9c11f38/45f15ea9-c4d4-4d60-94b9-9f0584a9e9fa to reach EXECUTE_COMPLETE
2024-11-27 14:25:19,844 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Change Set state is AVAILABLE, waiting 5(s)...
2024-11-27 14:25:24,975 INFO      [CloudFormation Stack akennedy-test-kingpin-1] Done updating template
```

image
